### PR TITLE
Store Prometheus labels as a slice instead of a map

### PR DESCRIPTION
**Pull request by swiatekm**
**Description:** 
Store labels for targets in the target allocator the same way Prometheus itself does - as a key+value list of type `labels.Labels` instead of a `model.Labelset` map. The former is what relabelling acts on, so we get to skip the conversion there, and building the list is faster as well. This makes looking up values of specific labels slower, but we don't really do that outside of determining the Node name.

I've modified the benchmark to actually drop half the targets during relabelling to make it more realistic.

Benchstat shows some nice gains:

```
goos: linux
goarch: amd64
pkg: github.com/open-telemetry/opentelemetry-operator/cmd/otel-allocator
cpu: AMD Ryzen 9 7950X3D 16-Core Processor          
                                                      │ bench_targets_main.txt │      bench_targets_branch.txt       │
                                                      │         sec/op         │   sec/op     vs base                │
ProcessTargets/least-weighted-32                                   78.24m ± 1%   42.75m ± 1%  -45.36% (p=0.000 n=10)
ProcessTargets/consistent-hashing-32                               78.86m ± 0%   42.72m ± 1%  -45.83% (p=0.000 n=10)
ProcessTargets/per-node-32                                         78.71m ± 0%   42.86m ± 1%  -45.55% (p=0.000 n=10)
ProcessTargetsWithRelabelConfig/per-node-32                        78.69m ± 1%   43.15m ± 1%  -45.17% (p=0.000 n=10)
ProcessTargetsWithRelabelConfig/least-weighted-32                  79.11m ± 1%   43.25m ± 1%  -45.33% (p=0.000 n=10)
ProcessTargetsWithRelabelConfig/consistent-hashing-32              79.05m ± 1%   42.86m ± 2%  -45.79% (p=0.000 n=10)
geomean                                                            78.78m        42.93m       -45.50%

                                                      │ bench_targets_main.txt │       bench_targets_branch.txt       │
                                                      │          B/op          │     B/op      vs base                │
ProcessTargets/least-weighted-32                                  45.14Mi ± 0%   40.41Mi ± 0%  -10.48% (p=0.000 n=10)
ProcessTargets/consistent-hashing-32                              45.15Mi ± 0%   40.41Mi ± 0%  -10.49% (p=0.000 n=10)
ProcessTargets/per-node-32                                        45.15Mi ± 0%   40.41Mi ± 0%  -10.49% (p=0.000 n=10)
ProcessTargetsWithRelabelConfig/per-node-32                       45.38Mi ± 0%   40.44Mi ± 0%  -10.88% (p=0.000 n=10)
ProcessTargetsWithRelabelConfig/least-weighted-32                 45.38Mi ± 0%   40.44Mi ± 0%  -10.88% (p=0.000 n=10)
ProcessTargetsWithRelabelConfig/consistent-hashing-32             45.38Mi ± 0%   40.44Mi ± 0%  -10.88% (p=0.000 n=10)
geomean                                                           45.26Mi        40.43Mi       -10.68%

                                                      │ bench_targets_main.txt │      bench_targets_branch.txt       │
                                                      │       allocs/op        │  allocs/op   vs base                │
ProcessTargets/least-weighted-32                                  112.36k ± 0%   52.05k ± 0%  -53.67% (p=0.000 n=10)
ProcessTargets/consistent-hashing-32                              112.38k ± 0%   52.05k ± 0%  -53.68% (p=0.000 n=10)
ProcessTargets/per-node-32                                        112.38k ± 0%   52.06k ± 0%  -53.68% (p=0.000 n=10)
ProcessTargetsWithRelabelConfig/per-node-32                       112.73k ± 0%   52.26k ± 0%  -53.64% (p=0.000 n=10)
ProcessTargetsWithRelabelConfig/least-weighted-32                 112.73k ± 0%   52.26k ± 0%  -53.64% (p=0.000 n=10)
ProcessTargetsWithRelabelConfig/consistent-hashing-32             112.74k ± 0%   52.26k ± 0%  -53.64% (p=0.000 n=10)
geomean                                                            112.6k        52.16k       -53.66%
```

### CODE REVIEW

File: cmd/otel-allocator/prehook/relabel.go
Comment (nicolastakashi) [inline cmd/otel-allocator/prehook/relabel.go:50]: I'm wondering if we could delete the job on the first if !keepTarget

Comment (swiatekm) [reply cmd/otel-allocator/prehook/relabel.go:50]: On the first !keepTarget, we break out of the loop, and then delete. So I think we're doing what you're suggesting?

Really grateful for your review on these kinds of changes by the way!

Comment (nicolastakashi) [reply cmd/otel-allocator/prehook/relabel.go:50]: I'm more than happy to help somehow I can 🙇🏽‍♂️ 
I was thinking about delete inside the if we break, does it makes sense?

**Comment by nicolastakashi** [comment]
Amazing work on that @swiatekm 👏🏽

**Review by nicolastakashi — APPROVED** [review]
LGTM

**Comment by nicolastakashi** [comment]
Hey @swiatekm anything I can do to help with this change?

**Comment by swiatekm** [comment]
@nicolastakashi I'm just waiting for another approver review.

**Pull request merged**
